### ISSUE TITLE
Can ultralytics, yolo inferencing scripts compilable with Nuitka?

### ISSUE THREAD

**Issue by WrldEngine** [ultralytics/ultralytics#21722]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I tried the command:
```
nuitka --standalone --onefile --include-package=cv2 --include-package=psutil  --include-package=ultralytics --lto=yes --jobs=8 main.py
```

And after 1 hour (btw large compile time), i get an error:
```
Nuitka-Options: Used command line options:
Nuitka-Options:   --standalone --onefile --include-package=cv2 --include-package=psutil --include-package=ultralytics --lto=yes --jobs=8 main.py
Nuitka: Starting Python compilation with:
Nuitka:   Version '2.7.12' on Python 3.12 (flavor 'Debian Python') commercial grade 'not installed'.
Nuitka-Plugins:WARNING: options-nanny: Module has plugin consideration: For matplotlib a GUI tooltkit selection might be needed use, use '--enable-plugin=tk-inter|PySide6|PySide2|PyQt6|PyQt5', or use '--enable-plugin=no-qt' for headless mode.
Nuitka-Plugins:WARNING: options-nanny: Module has parameter: Torch JIT is disabled by default in standalone mode, make a choice explicit with '--module-parameter=torch-disable-jit=yes|no'
Nuitka-Plugins:WARNING: anti-bloat: Undesirable import of 'pytest' in 'psutil.tests.test_scripts' (at '~/test/.venv/lib/python3.12/site-packages/psutil/tests/test_scripts.py:15') encountered. It may slow down compilation.
Nuitka-Plugins:WARNING:     Complex topic! More information can be found at [94mhttps://nuitka.net/info/unwanted-module.html
Nuitka-Plugins:anti-bloat: Not including '_json' automatically in order to avoid bloat, but this may cause: may slow down by using fallback implementation.
Nuitka-Plugins:anti-bloat: Not including 'PIL.ImageQt' automatically in order to avoid bloat, but this may cause: PIL will not be able to create Qt image objects.
Nuitka-Plugins:anti-bloat: Not including 'pandas.core._numba.extensions' automatically in order to avoid bloat, but this may cause: no numba acceleration.
Nuitka-Plugins:matplotlib: Using configuration file or default backend 'agg'.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.
Nuitka: Running C compilation via Scons.
Nuitka-Scons: Backend C compiler: gcc (gcc 13).
Nuitka-Scons: Backend C linking with 4203 files (no progress information available for this stage).
Nuitka-Scons: Running gcc -o ~/test/main.dist/main.bin -flto=8 -fuse-linker-plugin -fpartial-inlining -freorder-functions -O2 -s -z noexecstack -Xlinker -export-dynamic -rdynamic -Wl,-rpath=/usr/lib -Wl,-R,'$ORIGIN' -Wl,--disable-new-dtags -Wl,-b -Wl,binary -Wl,./__constants.bin -Wl,-b -Wl,elf64-x86-64 -Wl,-defsym -Wl,constant_bin_data=_binary_____constants_bin_start @"./@link_input.txt -L/usr/lib -ldl -lm /usr/lib/python3.12/config-3.12-x86_64-linux-gnu/libpython3.12-pic.a -lz -lm -lutil -lpthread -lexpat took 884.14 seconds
scons: *** [...] Build interrupted.
FATAL: Failed unexpectedly in Scons C backend compilation.
Nuitka:WARNING:     Complex topic! More information can be found at [94mhttps://nuitka.net/info/scons-backend-failure.html
Nuitka-Reports: Compilation crash report written to file 'nuitka-crash-report.xml'.
```

Maybe i should follow the instruction from WARNINGs, for example disabling torch jit?
Without ultralytics it compiles easly without problems

**Comment by UltralyticsAssistant** [comment]
👋 Hello @WrldEngine, thank you for your interest in Ultralytics 🚀! This is an automated response to help get you unblocked quickly. An Ultralytics engineer will also review your issue and assist soon.

We recommend a visit to the Docs for general guidance where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python usage: https://docs.ultralytics.com/usage/python/
- CLI usage: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example (MRE) to help us debug it:
- MRE guide: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our Tips for Best Training Results:
- Training tips: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord (real-time chat) 🎧: https://discord.com/invite/ultralytics
- Discourse (in-depth Q&A): https://community.ultralytics.com/
- Subreddit (community threads): https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify the issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

- Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit:
- Modes: https://docs.ultralytics.com/modes/
- Tasks: https://docs.ultralytics.com/tasks/

To help us investigate your Nuitka question more effectively, please:
- Share an MRE that reproduces the issue (a minimal main.py and exact Nuitka command you used).
- Include environment details: OS, Python and PyTorch versions, ultralytics version, GPU/CPU, and whether CUDA is installed.
- Provide the full Nuitka crash report (nuitka-crash-report.xml) and any logs showing the first error.
- Confirm the issue persists after upgrading ultralytics and dependencies as shown above.

Thanks for the detailed logs so far 🙏. Once we have the MRE and environment info, we can dive deeper.

**Comment by Y-T-G** [comment]
Can you explain what's the intended purpose of this? Are you trying to deploy your model?

**Comment by WrldEngine** [comment]
> Can you explain what's the intended purpose of this? Are you trying to deploy your model?

No, i want some inference logic, compile to binary executable file

**Comment by Y-T-G** [comment]
You can compile to binary by just writing the inference code in C++ or Rust. We have examples here:

https://github.com/ultralytics/ultralytics/tree/main/examples

**Comment by WrldEngine** [comment]
> You can compile to binary by just writing the inference code in C++ or Rust. We have examples here:
> 
> https://github.com/ultralytics/ultralytics/tree/main/examples

Yes, i know about it, but i faced with big codebase, that uses yolo detection, that's why i am asking, otherwise i would implement by myself

**Comment by glenn-jocher** [comment]
Got it—freezing Ultralytics + PyTorch into a single-file binary with Nuitka isn’t something we support or test, and link-time failures like this are common; you can try `--module-parameter=torch-disable-jit=yes`, but the reliable route is to export your model and run it from a small compiled runner instead—see the TorchScript export guide and options in the Export docs.

**Comment by WrldEngine** [comment]
> Got it—freezing Ultralytics + PyTorch into a single-file binary with Nuitka isn’t something we support or test, and link-time failures like this are common; you can try `--module-parameter=torch-disable-jit=yes`, but the reliable route is to export your model and run it from a small compiled runner instead—see the TorchScript export guide and options in the Export docs.

Thank you, but nuitka - just translates python code to C, and compiles with popular C compiler, thats why i didn't expect some pitfalls (and also could compile some custom libs with cuda functionality)

So, am i right that you suggest make some C/C++/Rust function that uses YOLO inference with ONNX, and with FFI call it from python? That means i reject your code, but use your model

**Comment by glenn-jocher** [comment]
Yes—that’s the idea: export your YOLO11 model and run it from a small compiled backend (ONNX Runtime/TensorRT/LibTorch), then call that from Python via FFI/subprocess so you don’t ship the Ultralytics/PyTorch stack at inference; for example: yolo export model=path/to/best.pt format=onnx nms=True. See the TorchScript export guide and options in the Export docs: https://docs.ultralytics.com/integrations/torchscript/ and https://docs.ultralytics.com/modes/export/.

**Comment by WrldEngine** [comment]
> Yes—that’s the idea: export your YOLO11 model and run it from a small compiled backend (ONNX Runtime/TensorRT/LibTorch), then call that from Python via FFI/subprocess so you don’t ship the Ultralytics/PyTorch stack at inference; for example: yolo export model=path/to/best.pt format=onnx nms=True. See the TorchScript export guide and options in the Export docs: https://docs.ultralytics.com/integrations/torchscript/ and https://docs.ultralytics.com/modes/export/.

thanks, but as i know onnx make inference slower, than pt, did you hear something about this?

**Comment by WrldEngine** [comment]
> Yes—that’s the idea: export your YOLO11 model and run it from a small compiled backend (ONNX Runtime/TensorRT/LibTorch), then call that from Python via FFI/subprocess so you don’t ship the Ultralytics/PyTorch stack at inference; for example: yolo export model=path/to/best.pt format=onnx nms=True. See the TorchScript export guide and options in the Export docs: https://docs.ultralytics.com/integrations/torchscript/ and https://docs.ultralytics.com/modes/export/.

Thanks, im not into ML, but can you suggest some tools for tensort inference in C++/Rust?

**Comment by glenn-jocher** [comment]
For a compiled runner, use one of these backends: C++ with ONNX Runtime for `.onnx`, NVIDIA TensorRT C++ API for `.engine`, or LibTorch for `.torchscript`; in Rust, look at the `onnxruntime` and `tch-rs` crates, with community TensorRT bindings also available. Export to the matching format and load it in your app—see the ONNX guide for export and deployment pointers in [ONNX export for YOLO11 models](https://docs.ultralytics.com/integrations/onnx/) and TensorRT specifics in the [NVIDIA Jetson guide](https://docs.ultralytics.com/guides/nvidia-jetson/).
